hollow/src/main/java/com/netflix/hollow/api/producer/HollowProducer.java

```diff
@@ -366,6 +387,26 @@ public boolean enablePrimaryProducer(boolean doEnable) {
         return (singleProducerEnforcer.isPrimary() == doEnable);
     }
 
+    /**
+     * Allows to pin a specific version, won't produce data, only modifies the pinned version file
+     *
+     */
+    public void pinVersion(long pinnedVersion) {
```

Comment:
Our tooling for performing a pin is operationally separate from our producer. There may be some good reasons for coupling the two, but for our use-case we'd really have a system that only had the `VersionPinner` and eschewed using a `HollowProducer` at all.

Said another way, if you wanted to build a microservice that was only responsible for pinning, it would heavy weight to have to instantiate a producer just to do it. Similarly, there's value in not having to rely on the producer that's producing your data states to do the actual pinning; one case could be if the producer somehow announced a bad state just prior to becoming unresponsive. Operationally you would like to pinback quickly while taking time on the producer remediation.

That said, I can see the motivation of wanting to consolidate the idea of pinning in both the `HollowConsumer` and `HollowProducer` interfaces.

Reply:
I thought about this too.

I agree that most of the times you'd like to have a `VersionPinner` to do the job without knowing `HollowProducer` at all.

I think it makes more sense to consolidate the pinning in the consumer rather than the producer. 

One thing I can think of is that if the `HollowProducer` knows about the `VersionPinner` and that a `pin` happened, you could potentially prevent the producer for executing more cycles while you fix any potential issue.  I know this can be done via tools to pin version and stop the producer (if needed)... so I was wondering if this could be a potential use case for adding the pinner to the producer. 

I think at the end, having a `VersionPinner` could be helpful for both cases... probably it shouldn't be part of the `HollowProducer` package at all, but the producer could know about it. Also, I didn't get into the details of listeners for the pinning mechanism on this initial proposal

Reply:
> One thing I can think of is that if the HollowProducer knows about the VersionPinner and that a pin happened, you could potentially prevent the producer for executing more cycles while you fix any potential issue. I know this can be done via tools to pin version and stop the producer (if needed)... so I was wondering if this could be a potential use case for adding the pinner to the producer.

That's also a good point. You might not always want to stop producer cycles when pinning. In our experience it depends on whether you're responding to a bug in the producer versus responding to bad data from your source-of-truth. In the case of bad source data you may want to keep running cycles because of the tooling (history, explorer) you get on published blobs.

Nevertheless, I can see that there might be potential value in having pinning visibility in the producer. As mentioned in the pinning section of the docs (and if you don't halt the producer when pinned) that you produce a bunch of bad delta states and you want the producer to patch over them by restoring from the pinned version. That could be justify tighter coupling between producer and pinner.